Update sentencepiece to newest version

emjotde opened the pull request:
### Description
* Updates sentencepiece to the newest version (removes dependency on protobuf)
* Enable SentencePiece compilation by default since there is no dependency in protobuf anymore.
* Seems that the current version has fewer possibilities to adjust training parameters and uses more RAM, but results for similar training seems to not diverge by much. 
* Note: despite moving from protobuf to protobuf-lite there seems to be binary-compatibility, i.e. spm models builts with old version work with this version. That is good news. 

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

emjotde commented:
@snukky Unsurprisingly the regression tests are failing if they compare text versions of vocabs. It seems the text files are written in different orders etc. I guess that would be expected after a longer break in updating/syncing. It does not look to me like there are any functionality problems.

emjotde commented:
Will address #659

emjotde commented:
These regression tests are failing:
```
  - tests/interface/input-tsv/test_tsv_train_create_vocabs.sh
  - tests/sentencepiece/test_create_joint_spm_vocab.sh
  - tests/sentencepiece/test_create_spm_vocab_lm.sh
  - tests/sentencepiece/test_create_spm_vocabs.sh
  - tests/sentencepiece/test_sentencepiece_max_lines.sh
  - tests/sentencepiece/test_sentencepiece_normalization.sh
```

snukky commented on src/common/config_parser.cpp:
```diff
@@ -372,7 +372,7 @@ void ConfigParser::addOptionsTraining(cli::CLIWrapper& cli) {
   cli.add<size_t>("--sentencepiece-max-lines",
       "Maximum lines to train SentencePiece vocabulary, selected with sampling from all data. "
```
If there is no sampling anymore, please update "selected with sampling from all data" in the help message.

emjotde replied:
I keep our sampling in there since it also concatenates the files from both translation directions. So there is more logic happening in the background.

emjotde replied:
As for the changes, I will ask Aaron to review them internally and do a separate PR when syncing changes back. I think most of the concerned protobuf which is now gone.

emjotde replied:
Leaving the description as is since we are still sampling.

snukky commented on src/3rd_party/CMakeLists.txt:
```diff
@@ -66,10 +66,6 @@ if(USE_SENTENCEPIECE)
   set(SPM_ENABLE_TCMALLOC ON CACHE BOOL "Enable TCMalloc if available.")
 
   if(USE_STATIC_LIBS)
-    message(WARNING "You are compiling SentencePiece binaries with -DUSE_STATIC_LIBS=on. \
```
Have you checked if it works now? Asking because the PR description does not mention this change.

emjotde replied:
Yes. works fine.

kpu replied:
😌💖

emjotde replied:
Added to "fixed" Changelog section

snukky commented on CMakeLists.txt:
```diff
@@ -236,6 +236,7 @@ endif()
 if(USE_ONNX)
   message(STATUS "Enabling experimental ONNX support")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_ONNX")
+  # updated sentencepiece likely requires this to find protobuf by itself
```
The comment says "likely". Should we leave a TODO here to verify this?

emjotde replied:
Yes, that what essentially meant to be a TODO, will update.

snukky commented on CHANGELOG.md:
```diff
@@ -53,6 +53,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Improved handling for receiving SIGTERM during training. By default, SIGTERM triggers 'save (now) and exit'. Prior to this fix, batch pre-fetching did not check for this sigal, potentially delaying exit considerably. It now pays attention to that. Also, the default behaviour of save-and-exit can now be disabled on the command line with --sigterm exit-immediately.
 
 ### Changed
+- Updated SentencePiece repository to version 8336bbd0c1cfba02a879afe625bf1ddaf7cd93c5 from https://github.com/google/sentencepiece. Enabled compilation by default since no dependency on protobuf anymore. Changed default value of --sentencepiece-max-lines to from 10000000 to 2000000 since apparently the new version doesn't sample automatically anymore. Not quite clear how that affects quality of  ...
```
"to from" -> "to" + I would suggest making a separate change including the second and third sentences.

emjotde replied:
Not quite sure I understand. Do you mean to have a separate bullet point for the 2. and 3. sentence? They do come from the same PR though?

snukky replied:
I mean that there are two changes: 1) updating SentencePiece, 2) changing the default value of an option; so maybe they deserve separate bullets. That's a minor comment.

emjotde replied:
Understood and changed accordingly.

snukky reviewed (commented):
Looks good, however our current fork has some changes

https://github.com/marian-nmt/sentencepiece/compare/6f4fe2815d42d0278da19c02014789643b8ad82b...dab002065a53a7622e2bfc2e972c76244b653ac3
https://github.com/marian-nmt/sentencepiece/commit/b92ee5f87cc534433193ed9adbaaaccdf322dcc5

that are not yet in the updated SentencePiece repo:

https://github.com/marian-nmt/sentencepiece/tree/8336bbd

Should those changes be transferred to the updated repo? I guess they might be relevant for the internal usage mainly/only.

snukky commented:
I updated regression tests: https://github.com/marian-nmt/marian-regression-tests/pull/68
I could simplify them a little as the vocab creation seems to be more deterministic now with the updated SentencePiece.

emjotde commented:
@snukky Thanks. I am now running two training with `examples/training-basics-sentencepiece` with old vs new to verify there is no major training-time regression.

emjotde commented:
Addressed review comments and update pointer to branch/commit with changed regression tests. All tests pass now. 

I am still running the contrastive trainings, but they seems to proceed at similar quality levels per iteration. Based on that I assume all is good, but will wait a couple of hours more.

emjotde commented:
- Updates sentencepiece to the newest version (removes dependency on protobuf)
- Enable SentencePiece compilation by default since there is no dependency in protobuf anymore.
